### ISSUE TITLE
Allow per-component logging configuration via environment variables

### ISSUE THREAD

**Issue by dfawley** [grpc/grpc-go#3937]
We already have component level logging functionality added here : [master/grpclog/component.go](https://github.com/grpc/grpc-go/blob/master/grpclog/component.go).

We want to add a new environment variable which will allow users to set different log levels for different components. For example GRPC_COMPONENT_LOG_LEVEL=dns:ERROR,authz:INFO

**Comment by easwars** [comment]
We did add the per component logging capability with https://github.com/grpc/grpc-go/blob/master/grpclog/component.go.

@dfawley : With this issue, did you envision having environment variables on a per component level where the verbosity and severity can be set?

**Comment by easwars** [comment]
This doesn't seem like a small enough issue which can be fixed as part of the fixit. But assigning the label anyways so that we can discuss it and make a call.

**Comment by ZhouyihaiDing** [comment]
Hi,
I am wondering if we can add a check for `GRPC_GO_LOG_SEVERITY_LEVEL` in [glogger](https://github.com/grpc/grpc-go/blob/master/grpclog/glogger/glogger.go).

**Comment by dfawley** [comment]
@ZhouyihaiDing this issue is unrelated.  Please send me a chat message to discuss or file a new issue, thanks!

**Comment by thefallentree** [comment]
what is the current way of showing these internal log messages? Even through code?

**Comment by easwars** [comment]
> what is the current way of showing these internal log messages? Even through code?

I'm not sure I understand your question. But we currently have two environment variables that control the logging level and the verbosity: `GRPC_GO_LOG_VERBOSITY_LEVEL` and `GRPC_GO_LOG_SEVERITY_LEVEL`.

**Comment by eshitachandwani** [comment]
The issue intends to add a environment variable to seperately specify logging levels for different components. Example : `GRPC_COMPONENT_LOG_LEVEL=dns:ERROR,authz:INFO`

**Comment by 110y** [comment]
@easwars 

Hi, I would like to try to implement this feature!

**Comment by easwars** [comment]
@110y I've assigned the issue to you. Please take a look at what we have and let's have a discussion here on the approach to take before you start implementing things. Thank you.

**Comment by 110y** [comment]
@easwars 

My plan is:

- Add a new environment variable `GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL` for per-component log level configuration
  - Format: `component1:LEVEL,component2:LEVEL` (e.g., `dns:ERROR,authz:INFO,xds:WARNING`)
      - Follow the existing `GRPC_GO_LOG_SEVERITY_LEVEL` for the case insensitivity.
  - I'm planning to add `SEVERITY_` to the environment variable name to make it align with existing `GRPC_GO_LOG_SEVERITY_LEVEL` variable.
- If `GRPC_GO_LOG_SEVERITY_LEVEL` and `GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL` are both set, the component-specific settings take precedence for that component.
- Parse `GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL` environment variable on `init` function in `grpclog/component.go` and cache the severity configurations as a package-level map variable.
- Modify logger functions (e.g. InfoDepth, InfoDepth etc.) on `grpclog.componentData` to make them respect the component-specific severity levels.

Please take a look 🙏

**Comment by easwars** [comment]
> Add a new environment variable GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL for per-component log level configuration 

Let's name the environment variable `GRPC_GO_COMPONENT_LOG_LEVEL` since adding `SEVERITY` in there makes it excessively long. But also because we can support both severity and verbosity with this same environment variable.

> If GRPC_GO_LOG_SEVERITY_LEVEL and GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL are both set, the component-specific settings take precedence for that component.

SG

> Parse GRPC_GO_COMPONENT_LOG_SEVERITY_LEVEL environment variable on init function in grpclog/component.go and cache the severity configurations as a package-level map variable.

SG

Not required initially. But we should make the changes here such that this can be extended for verbosity as follows `COMPONENT:LEVEL[:VERBOSITY],...`


> Modify logger functions (e.g. InfoDepth, InfoDepth etc.) on grpclog.componentData to make them respect the component-specific severity levels.

Please ensure that the required state (for example, the configured severity level) is passed to the `grpclog.componentData` at creation time, instead of every call to log requiring to access the global state.

**Comment by easwars** [comment]
A few more things to consider while implementing:
- Read the environment variable only once, at init time, instead of reading it every time a component logger is created
- The implementation of the component logger should be smart enough to not have to check for the configured severity level on all calls to Info/Warning/Error/Fatal etc. Instead it should be internally set up such that for example, if the configured severity level is `ERROR`, then any calls to Info/Warning should be no-ops. This can be achieved by delegating `Info(f)` calls to another type that will log at the correct depth if configured severity is `INFO`, and delegating to a no-op implementation when the configured severity is greater than `INFO`.

Please let me know if you have more questions.
